src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/DocumentSymbol/DocumentSymbolEndpoint.cs

```diff
@@ -0,0 +1,87 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT license. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Razor.Language;
+using Microsoft.AspNetCore.Razor.LanguageServer.Common;
+using Microsoft.AspNetCore.Razor.LanguageServer.EndpointContracts;
+using Microsoft.AspNetCore.Razor.LanguageServer.Extensions;
+using Microsoft.AspNetCore.Razor.LanguageServer.Protocol;
+using Microsoft.AspNetCore.Razor.PooledObjects;
+using Microsoft.CommonLanguageServerProtocol.Framework;
+using Microsoft.VisualStudio.LanguageServer.Protocol;
+
+namespace Microsoft.AspNetCore.Razor.LanguageServer.DocumentSymbol;
+
+[LanguageServerEndpoint(Methods.TextDocumentDocumentSymbolName)]
+internal class DocumentSymbolEndpoint : IRazorRequestHandler<DocumentSymbolParams, SymbolInformation[]>, ICapabilitiesProvider
+{
+    private readonly ClientNotifierServiceBase _languageServer;
+    private readonly IRazorDocumentMappingService _documentMappingService;
+
+    public DocumentSymbolEndpoint(ClientNotifierServiceBase languageServer, IRazorDocumentMappingService documentMappingService)
+    {
+        _languageServer = languageServer ?? throw new ArgumentNullException(nameof(languageServer));
+        _documentMappingService = documentMappingService ?? throw new ArgumentNullException(nameof(documentMappingService));
+    }
+
+    public bool MutatesSolutionState => false;
+
+    public void ApplyCapabilities(VSInternalServerCapabilities serverCapabilities, VSInternalClientCapabilities clientCapabilities)
+    {
+        serverCapabilities.DocumentSymbolProvider = new DocumentSymbolOptions()
+        {
+            WorkDoneProgress = false
+        };
+    }
+
+    public TextDocumentIdentifier GetTextDocumentIdentifier(DocumentSymbolParams request)
+        => request.TextDocument;
+
+    public async Task<SymbolInformation[]> HandleRequestAsync(DocumentSymbolParams request, RazorRequestContext context, CancellationToken cancellationToken)
+    {
+        if (request is null)
+        {
+            throw new ArgumentNullException(nameof(request));
+        }
+
+        var documentContext = context.GetRequiredDocumentContext();
+        var delegatedParams = new DelegatedDocumentSymbolParams(
+            new VersionedTextDocumentIdentifier()
+            {
+                Uri = documentContext.Uri,
+                Version = documentContext.Version,
+            }
```

Comment:
This needs ProjectContext I guess. 

Aside: it's a pity we have to do this, and can't just keep the whole object VS sent us. See Teams chat for more context